[rock] Switch to target attributes from serialization

fabianmcg opened the pull request:
This PR migrates rocMLIR to use GPU target attributes for compilation instead of the deprecated and removed `SerializeToHSA` pass.

Instead of `SerializeToHSA` the kernel pipeline now uses the passes `GpuROCDLAttachTarget` and `GpuModuleToBinaryPass`.

In addition, this PR also introduces a custom `gpu::TargetAttrInterface` for the
`ROCDLTargetAttr` attribute. This interface overrides the default behavior of:
`mlir::ROCDL::registerROCDLTargetInterfaceExternalModels`
By adding:
 - rocMLIR specific metadata in the binary.
 - Allowing to package device bitcode libs inside the rocMLIR binaries.
 - Use LLD API instead of calling the executable.

To register this interface one must call: `mlir::registerRocTarget`.

The idea of this custom interface is to minimize collisions with upstream, allowing to more customization on the rocMLIR side.

As consequence of this change, several parts of the infrastructure were updated to reflect the usage of `gpu.binary` ops intead of a GPU module with a binary string.

krzysz00 commented on external/llvm-project/mlir/include/mlir/Target/LLVM/ROCDL/Utils.h:
```diff
@@ -31,6 +31,15 @@ StringRef getROCMPath();
 class SerializeGPUModuleBase : public LLVM::ModuleToObject {
 public:
+  /// AMDGCN standard libraries.
+  typedef enum : uint32_t {
```
Could be an `enum class`?

fabianmcg replied:
Sure, I can make it an enum class, the reason I didn't make it an enum class is because it's already scoped inside the `SerializeGPUModuleBase`, but I'll take it out of the class and make it an enum class.

krzysz00 replied:
... Good point, it probably doesn't need to be an enum class because of the scoping.

I forget what upstream style on this sort of thing is

krzysz00 reviewed (commented):
This seems pretty reasonable overall

krzysz00 commented:
Re the PR, could you split the change into commits that only touch external/llvm-project, external/mlir-hal, and everything else?

fabianmcg commented:
> Re the PR, could you split the change into commits that only touch external/llvm-project, external/mlir-hal, and everything else?

Yes, I'll do it like that.

krzysz00 commented on mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp:
```diff
@@ -221,10 +221,14 @@ void rock::buildBackendPipeline(OpPassManager &pm,
-        options.triple, options.chip, options.features, options.optLevel,
-        options.suppressDiagnostic));
-    gpuPm.addPass(createRockCheckResidencyPass());
+    GpuROCDLAttachTargetOptions opts;
```
Note to future us: I wonder if it might be interesting to move this target attachment early in the pipeline

krzysz00 reviewed (commented):
I've got a few comments, haven't properly read the new translation widget yet

krzysz00 reviewed (commented):
Overall, the internal changes seem reasonable - I take it the copy-paste is somewhat inevitable here?

I'll ping @sjw36 for the MHAL changes

fabianmcg commented:
> I take it the copy-paste is somewhat inevitable here?

Are you referring to the LLVM upstream patches? or the Target attribute in rocMLIR?

manupak commented:
Only thing blocking the CI is fixed in https://github.com/ROCm/rocMLIR/commit/da8969573d2ad408c7ad129126679838d82d9350

Can you merge/rebase ?

manupak commented:
This PR has changes to two external repos and therefore it *should not* be squashed and merged.
Thus 22 commits is not something I d like to see in the tree.

Can you squash the commits to a sensible number while keeping the commits to each repo seperate.

fabianmcg commented:
I could potentially reduce it to 3 commits (llvm, mhal and roc), do we all agree on that?

manupak commented:
3 works for me; one each

fabianmcg committed:
[LLVM] Apply upstream LLVM patches for: SerializeToHSA to TargetAttr (1/3)

This patch applies a sequence of upstream LLVM patches required in the patch
series switching the compilation infrastructure from SerializeToHSA to `gpu`
target attributes.

The applied patches are:
- llvm pr #94918
- llvm pr #95541
- llvm pr #94991
- llvm pr #94998
- llvm pr #95456
- llvm pr #95292
- llvm pr #94908
- llvm pr #94910

fabianmcg committed:
[MHAL] Update MHAL to use gpu target attrs: SerializeToHSA to TargetAttr (2/3)

This patch updates the MHAL project to use the `gpu` target attributes
infrastructure. This is a patch in part of the series switching the compilation
infrastructure from SerializeToHSA to `gpu` target attributes.

This includes:
- Updating `PrefillPass` to work on GPU binaries.
- Updating `PackageTargetsPass` to work on GPU binaries.
- Updating `MHALToGPU` to create `gpu.binary` operations.
- Updating `mhal::TargetObject` to store an attribute instead of a string.
- Add the `DropMetadata` pass:
This pass drops all metadata from GPU binaries -e.g. property dictionary and
kernel metadata. This is required to avoid unregistered dialect errors
for attributes stored in the metadata like `mhal.prefill` but not parsable
by tools like `mlir-cpu-runner`.

fabianmcg committed:
[roc] Update roc to use gpu target attrs: SerializeToHSA to TargetAttr (3/3)

This patch updates the `rocMLIR` project to use the `gpu` target attributes
infrastructure. This is a patch in part of the series switching the compilation
infrastructure from SerializeToHSA to `gpu` target attributes.

This includes:
- Removing unnecessary headers and deprecated serialization passes.
- Adding new compilation passes to the rocMLIR kernel pipeline.
- Updating `CheckResidencyPass` to use `gpu.binary` ops.
This change also requires the pass to be run at the module level, as
`gpu.binary` ops are not isolated from above.
- Updating CAPI `mlirGetBinary` function to work on gpu binaries.
- Updating `rocmlir-lib` to work on `gpu.binary` ops.
- Adding missing gpu compilation registration calls to `xmir-runner`.
- Updating tests to work with the target attribute compilation mechanism.
- Updating the tunning driver to work on gpu binaries.
- Adding a target interface for the `ROCDLTargetAttr`:

This change introduces a custom `gpu::TargetAttrInterface` for the
`ROCDLTargetAttr` attribute. This interface overrides the default
behavior of:
`mlir::ROCDL::registerROCDLTargetInterfaceExternalModels`
By adding:
 - rocMLIR specific metadata in the binary.
 - Allowing to package device bitcode libs inside the rocMLI binaries.
 - Use LLD API instead of calling the executable.

To register this interface one must call:
`mlir::registerRocTarget`.

The idea of this custom interface is to minimize collisions with upstream.

krzysz00 reviewed (approved):
Approved here
